### ISSUE TITLE
Try Supervised Fine-Tuning on pseudo-QA-data

### ISSUE THREAD

**Issue by yk** [LAION-AI/Open-Assistant#48]
The first step in InstructGPT (https://openai.com/blog/instruction-following/) is supervised fine-tuning on human instruction data. Our website and bot are being created to collect this data. Meanwhile, we can already try out whether and how it's possible to fine-tune LLMs on such data, by substituting the not-yet-collected data for pseudo-data. One idea is to take a QA dataset (like squad or natural questions) and convert this into instruction-response pairs, then running the fine-tuning on top of that to get a feel for the dynamics of training.

**Comment by ekurtulus** [comment]
I can be of help for this.

**Comment by yk** [comment]
Amazing! Maybe it's best if you write down an initial roadmap here, like a step-by-step plan to reach the goal, then we can see better where potential problems might be!

**Comment by ekurtulus** [comment]
Here are the steps I am planning to take.

1. Find a good pretrained model which is not that large (I believe a mid-sized T5 would be a nice choice)
2. I will create two different variants of the SQUAD dataset, one with original labels and one with slightly corrupted labels (maybe back-translation ?)
3. I will train the reward model, basically a linear layer with output dims=2 on top of our pretrained model
4. Then, by using the reward model, I will try training the model with PPO

**Comment by bth5032** [comment]
Are we planning to use an encoder-decoder model like t5 or a decoder only model (or is this irrelevant at this stage)? I'm interested in understanding the choice of T5 vs a huggingface GPT2 implementation or OPT which has large pretrained models already available.

**Comment by yk** [comment]
> Are we planning to use an encoder-decoder model like t5 or a decoder only model (or is this irrelevant at this stage)? I'm interested in understanding the choice of T5 vs a huggingface GPT2 implementation or OPT which has large pretrained models already available.

architecturally, a decoder-only model is simpler, but an encoder-decoder system allows for explicit separation of input and output. ultimately, it will come down to the numbers which one we'd rather use. I feel a bit like decoder-only models shine in their simplicity and versatility, but I have no clue.

**Comment by ekurtulus** [comment]
The reason why I thought starting with T5 would be a good idea is that Flan-T5 outperforms OPT-IML. Also, once we have the training codebase available, we will be able to experiment with whatever architecture we want.

**Comment by bth5032** [comment]
> The reason why I thought starting with T5 would be a good idea is that Flan-T5 outperforms OPT-IML. Also, once we have the training codebase available, we will be able to experiment with whatever architecture we want.

Oh very interesting, wasn't aware of those papers. That's pretty counterintuitive (to me) that t5 vs gpt architecture would have that marked difference w.r.t. parameter efficiency, will have to look into it further! Thanks for explaining :)

**Comment by mrcabbage972** [comment]
@yk Please note that there are a few recent works on generating instructions data automatically. A couple of examples:

[SELF-INSTRUCT: Aligning Language Model
with Self Generated Instructions
](https://arxiv.org/pdf/2212.10560.pdf)

[Unnatural Instructions:
Tuning Language Models with (Almost) No Human Labor](https://arxiv.org/pdf/2212.09689.pdf)

The latter also released an instruction dataset of 240k samples [here](https://github.com/orhonovich/unnatural-instructions). I guess this could help us get started?

**Comment by yk** [comment]
@mrcabbage972 thank you very much for the pointers. do you think you could make a PR, create a new markdown file under `docs/research/` or something like this and start a collection of research works? The idea would be to have a place where we collect relevant papers, maybe a few tags for category, and a (short) description on what the paper could contribute to our efforts.

**Comment by mrcabbage972** [comment]
Sure, I can do that.

**Comment by mrcabbage972** [comment]
@yk Please see PR [here](https://github.com/LAION-AI/Open-Assistant/pull/217).

**Comment by sanagno** [comment]
Partially resolved in [PR](https://github.com/LAION-AI/Open-Assistant/pull/301).

**Comment by ekurtulus** [comment]
I am following up on your commit with the following:
- I am adding a script for creating pseudo-data for sanity checks and mock trainings
- Adding PolyLoss support which I heard that Palm team found to outperform CE
- Adding Sharpness Aware Minimization support
- Broader coverage of tasks
- Further optimization and extensions

**Comment by ekurtulus** [comment]
Plese see this [PR](https://github.com/LAION-AI/Open-Assistant/pull/576). I need a general feedback about the structure that I propose.

**Comment by lakshaykc** [comment]
@ekurtulus I'd like to help out on this issue. Any things you want me to look at? I guess once you create a new PR (replacing #576), we can go from there.

**Comment by ekurtulus** [comment]
> @ekurtulus I'd like to help out on this issue. Any things you want me to look at? I guess once you create a new PR (replacing #576), we can go from there.

Great ! @sanagno also works on this. I think at some point, we need to create a general roadmap of the features we want until the end of this month. @yk opinion on this ?

**Comment by sanagno** [comment]
@ekurtulus there are a few things/plans on in the discord server. I am not sure if I have seen you there :))

**Comment by ekurtulus** [comment]
> @ekurtulus there are a few things/plans on in the discord server. I am not sure if I have seen you there :))

Is it the ml-models channel ?

**Comment by sanagno** [comment]
Mostly yes, we can sync from there!

**Comment by lakshaykc** [comment]
> @ekurtulus there are a few things/plans on in the discord server. I am not sure if I have seen you there :))

Is that Open-Assistant discord or LAION discord? I just see the 'Lobby' channel in Open-Assistant discord.

**Comment by sanagno** [comment]
@lakshaykc what is your discord name? I can give you more access I believe

**Comment by lakshaykc** [comment]
> @lakshaykc what is your discord name? I can give you more access I believe

It is 'lkc'. Thanks.

**Comment by ekurtulus** [comment]
Please feel free to reopen this issue if the need for other tasks or models arise.

**Comment by lakshaykc** [comment]
Why was this closed? Is there a different issue related to this?

**Comment by sanagno** [comment]
The original purpose of the issue I think is served. See #619 for latest updates on the code. Once we get the human data we will continue on supervised fine-tuning on that. If you believe the discussion is better to continue here we can re-open this.

**Comment by lakshaykc** [comment]
Ok. I didn't see #619. Makes sense to just continue over there. We should update #200 as it mentions #48.
